upgrade_test.go

```diff
@@ -0,0 +1,213 @@
+	// git init
+	// git remote add origin <repository-url>
+	// git fetch origin <commit-hash>
+	// git checkout FETCH_HEAD
+
+	runCmdT(t, runArgs{
+		folder: folder,
+		cmd:    []string{"git", "init"},
+	})
+
+	runCmdT(t, runArgs{
+		folder: folder,
+		cmd:    []string{"git", "remote", "add", "origin", repo},
+	})
+
+	runCmdT(t, runArgs{
+		folder: folder,
+		cmd:    []string{"git", "fetch", "origin", sha},
+	})
+
+	runCmdT(t, runArgs{
+		folder: folder,
+		cmd:    []string{"git", "checkout", "FETCH_HEAD"},
+	})
+}
+
+func upgradeProvider(t *testing.T, folder string, targetVersion string, name string) {
+	cwd, err := os.Getwd()
+	require.NoError(t, err)
+
+	binPath := filepath.Join(cwd, "bin", "upgrade-provider")
+
+	runCmdT(t, runArgs{
+		folder: folder,
+		cmd: []string{
+			binPath,
+			"--kind", "provider",
+			"--target-version", targetVersion,
+			"--dry-run",
+			name,
+		},
+	})
+	require.NoError(t, err)
+}
+
+func upgradeBridge(t *testing.T, folder string, targetBridgeVersion string, name string) {
+	cwd, err := os.Getwd()
+	require.NoError(t, err)
+
+	binPath := filepath.Join(cwd, "bin", "upgrade-provider")
+
+	runCmdT(t, runArgs{
+		folder: folder,
+		cmd: []string{
+			binPath,
+			"--kind", "bridge",
+			"--target-bridge-version", targetBridgeVersion,
+			"--dry-run",
+			name,
+		},
+	})
+	require.NoError(t, err)
+}
+
+func runProviderUpgrade(t *testing.T, folder string, providerName string, sha string, targetVersion string) {
+	repo := fmt.Sprintf("https://github.com/%s.git", providerName)
+
+	runCheckout(t, folder, repo, sha)
+	upgradeProvider(t, folder, targetVersion, providerName)
+}
+
+func runBridgeUpgrade(t *testing.T, folder string, providerName string, sha string, targetBridgeVersion string) {
+	// use http instead of ssh to avoid ssh key issues
+	repo := fmt.Sprintf("https://github.com/%s.git", providerName)
+
+	runCheckout(t, folder, repo, sha)
+	upgradeBridge(t, folder, targetBridgeVersion, providerName)
+}
+
+func TestMain(m *testing.M) {
+	err := runCmd(runArgs{
+		folder: ".",
+		cmd:    []string{"make", "build"},
+	})
+
+	os.Setenv("GIT_COMMITTER_NAME", "Pulumi Bot")
+	os.Setenv("GIT_COMMITTER_EMAIL", "bot@pulumi.com")
+	os.Setenv("GIT_AUTHOR_NAME", "Pulumi Bot")
+	os.Setenv("GIT_AUTHOR_EMAIL", "bot@pulumi.com")
+
+	if err != nil {
+		fmt.Println("Error building the binary")
+		os.Exit(1)
+	}
+
+	os.Exit(m.Run())
+}
+
+func TestGCPMinorProviderUpgrade(t *testing.T) {
+	folder := tempDir(t, "pulumi-gcp")
+	version := "6.42.0"
+	runProviderUpgrade(t, folder, "pulumi/pulumi-gcp", "a9f15245447dd7806220110178325f00c964b278", version)
+
+	upgradeConfig, err := os.ReadFile(filepath.Join(folder, ".upgrade-config.yml"))
+	require.NoError(t, err)
+
+	require.Contains(t, string(upgradeConfig), fmt.Sprintf("current-upstream-version: %s", version))
+	// TODO: assert on upstream submodule sha
+}
+
+func TestGCPBridgeUpgrade(t *testing.T) {
+	folder := tempDir(t, "pulumi-gcp")
+	version := "bfbfac3b8a54f936f55f86cddf91a1b2184bc7b9"
```

Comment:
Interesting, at some point this will need to decouple from gcp but I like how you are pinning a commit so it is probably reasonable to start reusing that as a test case until the coupling issues actually arise.